File: NeetCode-150/Arrays and Hashing/Group Anagrams/solution.java
// sid 
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> ans = new ArrayList<>();
        Map<String, List<Integer>> map = new HashMap<>();
        
        for(int i = 0; i < strs.length; i++) {
            String str = strs[i];
            char[] ch = str.toCharArray();
            Arrays.sort(ch);
            String s = new String(ch);

            if(map.containsKey(s)) {
                List<Integer> ls = map.get(s);
                ls.add(i);

                map.put(s, ls);
            } else {
                map.put(s, new ArrayList<>(Arrays.asList(i)));
            }
        }

        for(List<Integer> list : map.values()) {
            List<String> ls = new ArrayList<>();
            for(int i = 0; i < list.size(); i++) {
                ls.add(strs[list.get(i)]);
            }
            ans.add(ls);
        }

        return ans;
    }
}

File: NeetCode-150/Arrays and Hashing/Group Anagrams/solution.cpp
// Hari

vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> res;
        
        // base case
        if (strs.size() == 1) {
            vector<string> val;
            val.push_back(strs[0]);
            res.push_back(val);
            return res;
        }
        
        // create map
        unordered_map<string, vector<string>> umap;
        
        for (int i = 0; i<strs.size(); i++) {
            string curr = strs[i];
            // sort the word
            sort(curr.begin(), curr.end());
            umap[curr].push_back(strs[i]);
        }
        
        for (auto it = umap.begin(); it != umap.end(); it++) {
            res.push_back(it -> second);
        }
        
        return res; 
    }

// Older solution (cleaner) 

vector<vector<string>> groupAnagrams(vector<string>& strs) {
        // fast
        ios_base::sync_with_stdio(false);
        // Logic: all anagrams of a type when sorted, result in same string
        
        vector<vector<string>> res;
        unordered_map<string, vector<string>> umap;
        
        for(auto it: strs){
            string temp = it;
            sort(temp.begin(), temp.end());
            umap[temp].push_back(it);        
        }
        
        for(auto it: umap){
            res.push_back(it.second);
        }
        
        return res;
    }
